### PR TITLE
dnsmasq and consul need to be better friends

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

**Comment by calston** [comment]
:+1: much wow..

File: manifests/consul_dns.pp [hunk 1]
```diff
@@ -61,6 +61,7 @@
   $encrypt            = undef,
   $bootstrap_expect   = undef,
   $ui                 = true,
+  $recursors          = [$::ipaddress_lo],
```
Comment (JayH5) [inline manifests/consul_dns.pp:71]: Docs pls

**Comment by JayH5** [comment]
+1, just the docs. Thanks for digging and figuring this out.
So, just to be clear:
- Client makes DNS request to 127.0.0.1:53 (Dnsmasq)
- Dnsmasq passes request to 127.0.0.1:8600 (Consul)
- Consul has hostname for Consul address, looks up address by recursing to 127.0.0.1:53 (Dnsmasq)
- Dnsmasq returns node address to Consul, Consul returns node hostname + address to Dnsmasq, Dnsmasq forwards that to client

Correct?

File: manifests/consul_dns.pp [hunk 2]
```diff
@@ -120,11 +123,16 @@
...
+    default   => $client_addr,
+  }
+
   # Dnsmasq
   # -------
   $dnsmasq_base_opts = {
     'cache-size' => '0',
-    'server'     => "/${domain}/${client_addr}#8600",
+    'server'     => "/${domain}/${dnsmasq_client_addr}#8600",
     'address'    => "/${dnsmasq_host_alias}/${advertise_addr}",
```
Comment (JayH5) [inline manifests/consul_dns.pp:143]: Oh also didn't we want to change this to `host-record`?

File: manifests/consul_dns.pp (hunk 1 above)
Comment (jerith) [inline manifests/consul_dns.pp:71]: Whoops, thanks for the reminder.

**Comment by jerith** [comment]
> - Client makes DNS request to 127.0.0.1:53 (Dnsmasq)
> - Dnsmasq passes request to 127.0.0.1:8600 (Consul)
> - Consul has hostname for Consul address, looks up address by recursing to 127.0.0.1:53 (Dnsmasq)
> - Dnsmasq returns node address to Consul, Consul returns node hostname + address to Dnsmasq, Dnsmasq forwards that to client

Yup, that's what it's doing. The trick to avoiding loops is to never send anything to consul that consul can't handle itself, so it only ever asks dnsmasq for things that dnsmasq isn't asking it for.

File: manifests/consul_dns.pp (hunk 2 above)
Comment (jerith) [inline manifests/consul_dns.pp:143]: We did. I'd prefer to do it in a new PR because I haven't tested the behaviour change here.

**Comment by JayH5** [comment]
:+1: (Tiny, optional nitpick "consul" -> "Consul" in your docs)

**Pull request merged**